### PR TITLE
Medical GUI - Show number of available treatment items in medical menu

### PR DESCRIPTION
**Pull request by amsteadrayle**
**When merged this pull request will:**
- Display the number of treatment items available on each treatment button in the medical menu 

### IMPORTANT

- [ ] If the contribution affects [the documentation](https://github.com/acemod/ACE3/tree/master/docs), please include your changes in this pull request so the documentation will appear on the [website](https://ace3.acemod.org/).
- [x] [Development Guidelines](https://ace3.acemod.org/wiki/development/) are read, understood and applied.
- [x] Title of this PR uses our standard template `Component - Add|Fix|Improve|Change|Make|Remove {changes}`.

### CODE REVIEW

**Comment by amsteadrayle** [comment]
This shows the items available in the medic's, patient's, and vehicle's inventories for each treatment, where applicable. If item sharing is disabled, the patient amount is hidden. When not in a vehicle, the vehicle amount is hidden.

The implementation could definitely be cleaned up, but this should be fully functional.

https://github.com/acemod/ACE3/assets/2516219/3ff30d7e-9e95-4651-843e-b5c8ddb19774

https://github.com/acemod/ACE3/assets/2516219/5059773a-0718-4945-aeed-5522f4eefd00

**Comment by amsteadrayle** [comment]
https://github.com/acemod/ACE3/blob/39ac1481482ffb0cd1ced4b23547e34db11b9c1f/addons/medical_gui/functions/fnc_countItems.sqf#L30

https://github.com/acemod/ACE3/blob/39ac1481482ffb0cd1ced4b23547e34db11b9c1f/addons/medical_gui/functions/fnc_countItems.sqf#L38

These both do the same exact thing, as far as I can tell. Is there any advantage to using the ACE common function over the second method? I included both so I could ask this question, and switching to either one should be fine.

**Comment by LinkIsGrim** [comment]
ACE implementation might be faster when dealing with large inventories. That's about it.

**Comment by amsteadrayle** [comment]
Alrighty. I'll go ahead and switch to that.

It's worth noting that the secondary text has priority over the primary text. I'm pretty certain there's enough room for both in English, but this may cause issues in other languages.
image

A quick scan of the string table shows Hungarian as one example:
image
image

**Comment by LinkIsGrim** [comment]
How will this work with sutures?

**Comment by amsteadrayle** [comment]
Good point, it doesn't acknowledge sutures. I can add a special case for stitching.

**Comment by amsteadrayle** [comment]
There is now an option to show this information as a tooltip instead, and that is the default.

https://github.com/acemod/ACE3/assets/2516219/04ce4c74-8169-4eef-946e-31d6d87e986f

**Comment by LinkIsGrim** [comment]
Tooltip is nice. I don't like the suture handling, though, but that's something for another PR.

**Comment by amsteadrayle** [comment]
What don't you like about the suture handling? I did just realize that I could be checking for `ACE_surgicalKit` in the items array instead of checking the display name.

**Comment by LinkIsGrim** [comment]
It's too specific (and I'm not a fan of checking the display name, but that's not the point). But a way to consume items as part of partial treatment was never implemented, so it's fine.

File: addons/medical_gui/functions/fnc_formatItemCounts.sqf [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+ * Example:
+ * [medicCount, patientCount, vehicleCount] call ace_medical_gui_fnc_formatItemCounts
+ *
+ * Public: No
+ */
+
+params ["_medicCount", "_patientCount", "_vehicleCount"];
+
+// Simple string padding function to control column width
+_fnc_padLeft = {
```
Comment (PabstMirror) [inline addons/medical_gui/functions/fnc_formatItemCounts.sqf:23]:
```suggestion
private _fnc_padLeft = {
```

File: addons/medical_gui/functions/fnc_formatItemCounts.sqf [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+    params ["_string", "_length"];
+    if (count _string >= _length) exitWith {_string};
+    private _full = "    " + _string;
+    _Full select [(count _full)-_length, _length];
+};
+
+_fnc_formatOnButton = {
```
Comment (PabstMirror) [inline addons/medical_gui/functions/fnc_formatItemCounts.sqf:30]:
```suggestion
private _fnc_formatOnButton = {
```

File: addons/medical_gui/functions/fnc_formatItemCounts.sqf [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+    };
+
+    if (!isNil {_vehicleCount}) then {
+        _countStrings pushBack ([str _vehicleCount, _columnWidth] call _fnc_padLeft);
+    };
+
+    _countStrings joinString "|"
+};
+
+_fnc_formatInTooltip = {
```
Comment (PabstMirror) [inline addons/medical_gui/functions/fnc_formatItemCounts.sqf:50]:
```suggestion
private _fnc_formatInTooltip = {
```

File: addons/medical_gui/functions/fnc_countTreatmentItems.sqf [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+/*
+ * Author: AmsteadRayle
+ * Counts how many of the given items are present between the medic and patient.
+ * If medic or patient are in a vehicle then vehicle's inventory will also be checked.
+ *
+ * Arguments:
+ * 0: Items <ARRAY>
+ *
+ * Return Value:
+ * Counts <ARRAY>
```
Comment (PabstMirror) [inline addons/medical_gui/functions/fnc_countTreatmentItems.sqf:11]:
```suggestion
 * Counts (can be nil) <ARRAY>
```

**Comment by LinkIsGrim** [comment]
On second thought, this should be tooltip only IMO (both to facilitate not breaking menu text and to prevent information overload).

**Comment by amsteadrayle** [comment]
I disagree with the button text being information overload, especially with it not being the default way to display the information.

But I do agree that it's not great to include something that could just make parts of the menu unreadable in certain languages. It would also make the formatting code way simpler to not have to handle the button text.

File: addons/medical_gui/functions/fnc_updateActions.sqf [hunk 5]
```diff
@@ -38,18 +38,33 @@ if (_showTriage) exitWith {
...
+    _x params ["_displayName", "_category", "_condition", "_statement", "_items"];
 
     // Check action category and condition
     if (_category == _selectedCategory && {call _condition}) then {
         private _ctrl = if (_shownIndex >= count _actionButons) then {
             _actionButons pushBack (_display ctrlCreate ["ACE_Medical_Menu_ActionButton", -1, _group]);
         };
+
+        private _countText = "";
+        if ((GVAR(showTreatmentItemCount) != 0) && (count _items != 0)) then {
```
Comment (johnb432) [inline addons/medical_gui/functions/fnc_updateActions.sqf:50]:
```suggestion
        if ((GVAR(showTreatmentItemCount) != 0) && (_items isNotEqualTo [])) then {
```

File: addons/medical_gui/functions/fnc_formatItemCounts.sqf [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+
+    private _countStrings = [[str _medicCount, _columnWidth] call _fnc_padLeft];
+
+    // If sharing disabled, don't show at all
+    // If sharing enabled but no target, show null column
+    if (EGVAR(medical_treatment,allowSharedEquipment) != 2) then {
+        _countStrings pushBack (if (isNil {_patientCount}) then {_nullString} else {[str _patientCount, _columnWidth] call _fnc_padLeft});
+    };
+
+    if (!isNil {_vehicleCount}) then {
```
Comment (johnb432) [inline addons/medical_gui/functions/fnc_formatItemCounts.sqf:43]:
```suggestion
    if (!isNil "_vehicleCount") then {
```

File: addons/medical_gui/functions/fnc_formatItemCounts.sqf [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+    params ["_medicCount", "_patientCount", "_vehicleCount"];
+    private _nullString = "——"; // Two em dashes to fill space with single horizontal bar
+    private _columnWidth = 2;
+
+    private _countStrings = [[str _medicCount, _columnWidth] call _fnc_padLeft];
+
+    // If sharing disabled, don't show at all
+    // If sharing enabled but no target, show null column
+    if (EGVAR(medical_treatment,allowSharedEquipment) != 2) then {
+        _countStrings pushBack (if (isNil {_patientCount}) then {_nullString} else {[str _patientCount, _columnWidth] call _fnc_padLeft});
```
Comment (johnb432) [inline addons/medical_gui/functions/fnc_formatItemCounts.sqf:40]:
```suggestion
        _countStrings pushBack (if (isNil "_patientCount") then {_nullString} else {[str _patientCount, _columnWidth] call _fnc_padLeft});
```

File: addons/medical_gui/functions/fnc_formatItemCounts.sqf [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+    private _columnWidth = 3;
+
+    private _countStrings = [format ["%1 %2", _medicCount, localize LSTRING(TreatmentItemCount_Medic)]];
+
+    if ((EGVAR(medical_treatment,allowSharedEquipment) != 2) && (!isNil {_patientCount})) then {
+        _countStrings pushBack format ["%1 %2", _patientCount, localize LSTRING(TreatmentItemCount_Patient)];
+    };
+
+    if (!isNil {_vehicleCount}) then {
+        _countStrings pushBack format ["%1 %2", _vehicleCount, localize LSTRING(TreatmentItemCount_Vehicle)];
```
Comment (johnb432) [inline addons/medical_gui/functions/fnc_formatItemCounts.sqf:61]:
```suggestion
    private _countStrings = [format ["%1 %2", _medicCount, LLSTRING(TreatmentItemCount_Medic)]];

    if ((EGVAR(medical_treatment,allowSharedEquipment) != 2) && {!isNil "_patientCount"}) then {
        _countStrings pushBack format ["%1 %2", _patientCount, LLSTRING(TreatmentItemCount_Patient)];
    };

    if (!isNil "_vehicleCount") then {
        _countStrings pushBack format ["%1 %2", _vehicleCount, LLSTRING(TreatmentItemCount_Vehicle)];
```

File: addons/medical_gui/functions/fnc_formatItemCounts.sqf [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+    if (!isNil {_vehicleCount}) then {
+        _countStrings pushBack ([str _vehicleCount, _columnWidth] call _fnc_padLeft);
+    };
+
+    _countStrings joinString "|"
+};
+
+private _fnc_formatInTooltip = {
+    params ["_medicCount", "_patientCount", "_vehicleCount"];
+    private _columnWidth = 3;
```
Comment (johnb432) [inline addons/medical_gui/functions/fnc_formatItemCounts.sqf:52]:
```suggestion
```

File: addons/medical_gui/functions/fnc_formatItemCounts.sqf [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+ */
+
+params ["_medicCount", "_patientCount", "_vehicleCount"];
+
+// Simple string padding function to control column width
+private _fnc_padLeft = {
+    params ["_string", "_length"];
+    if (count _string >= _length) exitWith {_string};
+    private _full = "    " + _string;
+    _Full select [(count _full)-_length, _length];
```
Comment (johnb432) [inline addons/medical_gui/functions/fnc_formatItemCounts.sqf:27]:
```suggestion
    _full select [(count _full)-_length, _length];
```

File: addons/medical_gui/functions/fnc_countTreatmentItems.sqf [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+params ["_items"];
+
+private _medicCount = 0;
+private _patientCount = nil;
+private _vehicleCount = nil;
+
+// Medic
+{
+    private _item = _x;
+    _medicCount = _medicCount + ([ACE_player, _item] call EFUNC(common,getCountOfItem));
```
Comment (johnb432) [inline addons/medical_gui/functions/fnc_countTreatmentItems.sqf:28]:
```suggestion
    _medicCount = _medicCount + ([ACE_player, _x] call EFUNC(common,getCountOfItem));
```
Wasteful and doesn't contribute to code readability, in my opinion.

File: addons/medical_gui/functions/fnc_countTreatmentItems.sqf [hunk 12]
```diff
@@ -0,0 +1,55 @@
...
+} forEach _items;
+
+// Patient
+if (ACE_player != GVAR(target)) then {
+    _patientCount = 0;
+    {
+        private _item = _x;
+        _patientCount = _patientCount + ([GVAR(target), _item] call EFUNC(common,getCountOfItem));
```
Comment (johnb432) [inline addons/medical_gui/functions/fnc_countTreatmentItems.sqf:36]:
```suggestion
        _patientCount = _patientCount + ([GVAR(target), _x] call EFUNC(common,getCountOfItem));
```

Same as above

**Comment by LinkIsGrim** [comment]
> On second thought, this should be tooltip only IMO (both to facilitate not breaking menu text and to prevent information overload).

Can get this into 3.16.0 still if this and merge conflicts are handled. Drop the setting, always on.

File: addons/medical_gui/stringtable.xml [hunk 13]
```diff
@@ -1349,5 +1349,14 @@
...
             <English>R</English>
         </Key>
+        <Key ID="STR_ACE_Medical_GUI_TreatmentItemCount_Medic">
+            <English>in your inventory</English>
+        </Key>
+        <Key ID="STR_ACE_Medical_GUI_TreatmentItemCount_Patient">
+            <English>in patient inventory</English>
+        </Key>
+        <Key ID="STR_ACE_Medical_GUI_TreatmentItemCount_Vehicle">
+            <English>in vehicle inventory</English>
```
Comment (LinkIsGrim) [inline addons/medical_gui/stringtable.xml:1359]:
```suggestion
        <Key ID="STR_ACE_Medical_GUI_TreatmentItemCount_Patient">
            <English>in patient's inventory</English>
        </Key>
        <Key ID="STR_ACE_Medical_GUI_TreatmentItemCount_Vehicle">
            <English>in vehicle's inventory</English>
```

**Review by LinkIsGrim — APPROVED** [review]

File: addons/medical_gui/functions/fnc_formatItemCounts.sqf [hunk 14]
```diff
@@ -0,0 +1,32 @@
+#include "..\script_component.hpp"
+/*
+ * Author: AmsteadRayle
+ * Format item counts to be shown either on the treatment button or in the tooltip.
```
Comment (LinkIsGrim) [inline addons/medical_gui/functions/fnc_formatItemCounts.sqf:4]:
```suggestion
 * Format item counts to be shown either in the tooltip.
```

**Pull request merged**